homeassistant/components/tuya.py

```diff
@@ -0,0 +1,166 @@
+"""
+Support for Tuya Smart devices.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/tuya/
+"""
+from datetime import timedelta
+import logging
+import voluptuous as vol
+
+from homeassistant.core import callback
+import homeassistant.helpers.config_validation as cv
+from homeassistant.const import (CONF_USERNAME, CONF_PASSWORD)
+from homeassistant.helpers import discovery
+from homeassistant.helpers.dispatcher import (
+    dispatcher_send, async_dispatcher_connect)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import track_time_interval
+
+REQUIREMENTS = ['tuyapy==0.1.1']
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_COUNTRYCODE = 'country_code'
+
+DOMAIN = 'tuya'
+DATA_TUYA = 'data_tuya'
+
+SIGNAL_DELETE_ENTITY = 'tuya_delete'
+SIGNAL_UPDATE_ENTITY = 'tuya_update'
+
+SERVICE_FORCE_UPDATE = 'force_update'
+SERVICE_PULL_DEVICES = 'pull_devices'
+
+TUYA_TYPE_TO_HA = {
+    'switch': 'switch'
+}
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_COUNTRYCODE): cv.string
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up Tuya Component."""
+    from tuyapy import TuyaApi
+
+    tuya = TuyaApi()
+    username = config[DOMAIN][CONF_USERNAME]
+    password = config[DOMAIN][CONF_PASSWORD]
+    country_code = config[DOMAIN][CONF_COUNTRYCODE]
+
+    hass.data[DATA_TUYA] = tuya
+    tuya.init(username, password, country_code)
+    hass.data[DOMAIN] = {
+        'entities': {}
+    }
+
+    def load_devices(device_list):
+        """Load new devices by device_list."""
+        device_type_list = {}
+        for device in device_list:
+            dev_type = device.device_type()
+            if (dev_type in TUYA_TYPE_TO_HA and
+                    device.object_id() not in hass.data[DOMAIN]['entities']):
+                ha_type = TUYA_TYPE_TO_HA[dev_type]
+                if ha_type not in device_type_list:
+                    device_type_list[ha_type] = []
+                device_type_list[ha_type].append(device.object_id())
```

Comment:
We should probably populate the entities dict with the device id here too so that we don't load the same device again if `load_devices`would be called again before the entity has been added to home assistant.

```py
hass.data[DOMAIN]['entities'][dev_id] = None
```